### ISSUE TITLE
"command -v 'php' || which 'php' || type -p 'php'" failed.

### ISSUE THREAD

**Issue by szwss** [lorisleiva/laravel-deployer#119]
I have encountered this problem in use, what is the problem, thank you, the following is the error message:
```
✈︎ Deploying master on 39.108.110.255 with basic strategy
➤ Executing task deploy:prepare
✔oot@39.108.110.255's password:
✔ Executing task deploy:lock
✔ Executing task deploy:release
✔ Executing task deploy:update_code
✔ Executing task deploy:shared
➤ Executing task deploy:vendors
✔ Executing task deploy:failed
✔ Executing task deploy:unlock

In Client.php line 99:

  The command "command -v 'php' || which 'php' || type -p 'php'" fail
  ed.

  Exit Code: 1 (General error)

  Host Name: 39.108.110.255

  ================


deploy [-p|--parallel] [-l|--limit LIMIT] [--no-hooks] [--log LOG] [--roles ROLES] [--hosts HOSTS] [-o|--option OPTION] [--] [<stage>]


   Symfony\Component\Process\Exception\ProcessFailedException  : The command "vendor/bin/dep --file=vendor/lorisleiva/laravel-deployer/.build/deploy.php deploy " failed.

Exit Code: 1(General error)

Working directory: /home/vagrant/code/jtd/logistics

Output:
================


Error Output:
================


  at /home/vagrant/code/jtd/logistics/vendor/symfony/process/Process.php:256
    252|      */
    253|     public function mustRun(callable $callback = null, array $env = [])
    254|     {
    255|         if (0 !== $this->run($callback, $env)) {
  > 256|             throw new ProcessFailedException($this);
    257|         }
    258|
    259|         return $this;
    260|     }

  Exception trace:

  1   Symfony\Component\Process\Process::mustRun(Object(Closure))
      /home/vagrant/code/jtd/logistics/vendor/lorisleiva/laravel-deployer/src/LaravelDeployer/Commands/BaseCommand.php:110

  2   Lorisleiva\LaravelDeployer\Commands\BaseCommand::process("vendor/bin/dep --file=vendor/lorisleiva/laravel-deployer/.build/deploy.php deploy ")
      /home/vagrant/code/jtd/logistics/vendor/lorisleiva/laravel-deployer/src/LaravelDeployer/Commands/BaseCommand.php:58

  Please use the argument -v to see more details.
```
The following is the contents of my configuration file:
```
<?php

return [

    /*
    |--------------------------------------------------------------------------
    | Default deployment strategy
    |--------------------------------------------------------------------------
    |
    | This option defines which deployment strategy to use by default on all
    | of your hosts. Laravel Deployer provides some strategies out-of-box
    | for you to choose from explained in detail in the documentation.
    |
    | Supported: 'basic', 'firstdeploy', 'local', 'pull'.
    |
    */

    'default' => 'basic',

    /*
    |--------------------------------------------------------------------------
    | Custom deployment strategies
    |--------------------------------------------------------------------------
    |
    | Here, you can easily set up new custom strategies as a list of tasks.
    | Any key of this array are supported in the `default` option above.
    | Any key matching Laravel Deployer's strategies overrides them.
    |
    */

    'strategies' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Hooks
    |--------------------------------------------------------------------------
    |
    | Hooks let you customize your deployments conveniently by pushing tasks 
    | into strategic places of your deployment flow. Each of the official
    | strategies invoke hooks in different ways to implement their logic.
    |
    */

    'hooks' => [
        // Right before we start deploying.
        'start' => [
            //
        ],
        
        // Code and composer vendors are ready but nothing is built.
        'build' => [
            //
        ],
        
        // Deployment is done but not live yet (before symlink)
        'ready' => [
            'artisan:storage:link',
            'artisan:view:clear',
            'artisan:cache:clear',
            'artisan:config:cache',
            'artisan:migrate',
        ],
        
        // Deployment is done and live
        'done' => [
            'fpm:reload',
        ],
        
        // Deployment succeeded.
        'success' => [
            //
        ],
        
        // Deployment failed.
        'fail' => [
            //
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Deployment options
    |--------------------------------------------------------------------------
    |
    | Options follow a simple key/value structure and are used within tasks
    | to make them more configurable and reusable. You can use options to
    | configure existing tasks or to use within your own custom tasks.
    |
    */

    'options' => [
        'application' => env('APP_NAME', 'Laravel'),
        'repository' => 'git@github.com:szwss/laravel-template.git',
        'php_fpm_service' => 'php7.2-fpm',
    ],

    /*
    |--------------------------------------------------------------------------
    | Hosts
    |--------------------------------------------------------------------------
    |
    | Here, you can define any domain or subdomain you want to deploy to.
    | You can provide them with roles and stages to filter them during
    | deployment. Read more about how to configure them in the docs.
    |
    */

    'hosts' => [
        '39.108.110.255' => [
            'deploy_path' => '/data/wwwroot/logistics.test',
            'user' => 'root',
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Localhost
    |--------------------------------------------------------------------------
    |
    | This localhost option give you the ability to deploy directly on your
    | local machine, without needing any SSH connection. You can use the
    | same configurations used by hosts to configure your localhost.
    |
    */

    'localhost' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Include additional Deployer recipes
    |--------------------------------------------------------------------------
    |
    | Here, you can add any third party recipes to provide additional tasks, 
    | options and strategies. Therefore, it also allows you to create and
    | include your own recipes to define more complex deployment flows.
    |
    */

    'include' => [
        //
    ],

    /*
    |--------------------------------------------------------------------------
    | Use a custom Deployer file
    |--------------------------------------------------------------------------
    |
    | If you know what you are doing and want to take complete control over
    | Deployer's file, you can provide its path here. Note that, without
    | this configuration file, the root's deployer file will be used.
    |
    */

    'custom_deployer_file' => false,

];
```
I am running these 3 commands on the server is normal
```
Last login: Fri Jun  7 19:15:09 2019 from 27.17.85.145
root@Homestead:~# which php
/usr/local/php/bin/php
root@Homestead:~#  type -p 'php'
/usr/local/php/bin/php
root@Homestead:~# command -v 'php'
/usr/local/php/bin/php
```

**Comment by szwss** [comment]
HI, is there anyone? What is this problem, who can help me, where the code is not found, and there are no valuable tips....

**Comment by SerendipityNL** [comment]
I'm having a similar problem but with `npm`. The command `which npm` fails, while it's perfectly fine when I enter it manually. If you need any info let me know!

**Comment by lorisleiva** [comment]
Hi there 👋 

Have you tried forcing a certain binary by setting the `bin/php` and/or `bin/npm` options?

**Comment by moreishi** [comment]
Does your server have composer installed?

**Comment by lorisleiva** [comment]
Closing due to inactivity. Feel free to open a new issue if the problem persists.
